atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -2384,13 +2386,22 @@ public void useTable(TableReference tableRef, ConstraintCheckable table) {
     /** The similarly-named-and-intentioned useTable method is only called on writes.
      *  This one is more comprehensive and covers read paths as well
      * (necessary because we wish to get the sweep strategies of tables in read-only transactions)
+     *
+     * A table can be involved in a transaction, even if there were no reads done on it, see #markTableInvolved.
      */
     private void markTableAsInvolvedInThisTransaction(TableReference tableRef) {
         involvedTables.add(tableRef);
     }
 
+    private boolean hasAnyInvolvedTables() {
+        return !involvedTables.isEmpty();
+    }
+
     private boolean validationNecessaryForInvolvedTablesOnCommit() {
-        return involvedTables.stream().anyMatch(this::isValidationNecessaryOnCommit);
+        boolean anyTableRequiresImmutableTimestampLocking = involvedTables.stream().anyMatch(
+                this::requiresImmutableTimestampLocking);
+        boolean needsToValidate = !validateLocksOnReads || !hasReads();
+        return anyTableRequiresImmutableTimestampLocking && needsToValidate;
```

Comment:
So we validate if there's an interesting table AND either there were no reads (because lock watches), or validate locks on reads was false.

There are potentially cases where this might behave differently from the original:

```
user reads table A, checks lock
<loses immutable timestamp lock>
user reads table B via notifications, marked B as used
txn.commit()
```

This would previously throw, but now reads the value from the cache, which is fine - the goal of the immutable timestamp lock is resilience against a delete clearing the value we want out from under us by Thorough Sweep.

Is the above understanding correct? May be good to have a quick check with @gmaretic here

Reply:
Where would it throw? Let me clarify my understanding: I understand your timeline as "table A read checks lock and still has it, but lock expires just afterwards". With that:

Before my change the behavior would be to not throw: when A does the read, it checks the lock, then user does some reads from a cache (AtlasDB is not aware of those reads), and then user calls commit. Then ``#hasReads`` is true (because of the A read) so we recheck PreCommitCondition, but #validationNecessaryForInvolvedTablesOnCommit returns false, because #isValidationNecessaryOnCommit checks ``!validateLocksOnReads``. And ``validateLocksOnReads`` is a preeety deep rabbit hole, but if you follow it it ultimately is set from ``TransactionManagers#validateLocksOnReads`` where the default is true.

After my change, hasReads is true, so again we check PreCommitCondition, and then anyTableRequiresImmutableTimestampLocking is true, because I'm assuming that's what your example does, ``!validateLocksOnReads`` is false, but ``!hasReads()`` is true, so we do it. I guess I could add a recheck for ``!hasReads() && hasAnyInvolvedTables()`` just to be nice?

Ultimately, I think what I care about most is the PreCommitCondition check: the immutable timestamp lock, as you say, is just for TargetedSweep to not nuke the cells under us, but we kinda don't care (because we have them in-memory!). So if you think this is just too complicated I can ignore, but I think I'd rather have this behave roughly the same as the case of only cache reads (the only difference is that the immutable lock check will be checked at commit time).

Reply:
In my post when I meant "original" I meant "before lock watches even exist", so the operation is reading A and then reading B - i.e. my concern is about the difference between

```
user reads table A, checks lock
<loses immutable timestamp lock>
user reads table B without notifications (because they don't exist), checks lock, fails
txn.commit()
```
and
```
user reads table A, checks lock
<loses immutable timestamp lock>
user reads table B via notifications, marked B as used
txn.commit()
```

Yep, I think the thing I wanted to verify was that this is _only_ for sweep protection, which I believe is correct - if that's the case then I'm pretty sure this is fine

Reply:
Yep, discussed offline. I think we don't need to make it this strict, as long as there's a check during the transaction's lifetime, I think that's fine by me.

Reply:
Just as a note for future awareness, we had a number of bad legacy users which did goofy things like read data read by a transaction (that was not the thing returned) outside of the scope of that transaction. Among other things, this is why we had to have the whole `checkLocksOnReads` true by default thing (and so "check during the transaction's lifetime" might actually not be sufficient _in general_ - though I don't have concrete examples to hand), but within the parameters of our usage is definitely fine.